pkg/ingester/user_metrics_metadata.go

```diff
@@ -86,15 +87,31 @@ func (mm *userMetricsMetadata) purge(deadline time.Time) {
 	mm.metrics.memMetadataRemovedTotal.WithLabelValues(mm.userID).Add(float64(deleted))
 }
 
-func (mm *userMetricsMetadata) toClientMetadata() []*mimirpb.MetricMetadata {
+func (mm *userMetricsMetadata) toClientMetadata(req *client.MetricsMetadataRequest) []*mimirpb.MetricMetadata {
 	mm.mtx.RLock()
 	defer mm.mtx.RUnlock()
+	if req == nil {
+		req = &client.MetricsMetadataRequest{Limit: -1, LimitPerMetric: -1, Metric: ""}
+	}
 	r := make([]*mimirpb.MetricMetadata, 0, len(mm.metricToMetadata))
```

Comment:
Should we preallocate this with the smaller of `mm.metricToMetadata` and `numMetrics`?